Bug 1862120: LocalVolumeSet should delay provisioning after discovery so it won't race with other provisioners.

rohantmp opened the pull request.

rohantmp commented:
/bugzilla refresh

gnufied commented on pkg/diskmaker/controllers/lvset/device_age.go:
```diff
@@ -0,0 +1,31 @@
+
+// getFirstObserved returns the age that the key was first observed and records it if
+// it was first observed now
+func getFirstObserved(deviceAgeMap *map[string]time.Time, mux *sync.Mutex, key string, getTime func() time.Time) time.Time {
```
I think this might be better written as a new type with mutex and map as properties of the new type. Something like:

```
type DeviceObservationAge struct{
  deviceAgeMap map[string]string
  mu *sync.Mutex
}
```

rohantmp replied:
done

rohantmp commented:
verified after changes: 
```
Events:
  Type    Reason               Age                From                               Message
  ----    ------               ----               ----                               -------
  Normal  DiscoveredNewDevice  62s (x6 over 62s)  localvolumeset-symlink-controller  minikube - found possible matching disk, waiting 1m0s to claim
  Normal  FoundMatchingDisk    2s (x3 over 2s)    localvolumeset-symlink-controller  minikube - symlinking matching disk
```

gnufied commented on pkg/diskmaker/controllers/lvset/device_age.go:
```diff
@@ -0,0 +1,52 @@
+type ageMap struct {
+	ageMap  map[string]time.Time
+	mux     sync.Mutex
+	getTime *func() time.Time
```
Can we rename this function to `currentTime` which is what actually is. Also add a comment here that this is just for test purpose?

gnufied commented on pkg/diskmaker/controllers/lvset/reconcile.go:
```diff
@@ -136,20 +136,33 @@ func (r *ReconcileLocalVolumeSet) Reconcile(request reconcile.Request) (reconcil
 	for _, blockDevice := range blockDevices {
 
+		// get/set firstObserved
+		firstObserved := r.deviceAgeMap.getFirstObserved(blockDevice.KName)
```
it might be cleaner to rename this function as  - `availableAt` so as this function reports time at which device will become available here.

rohantmp replied:
hmm, I tried to keep `ageMap` generic. this would compromise that

gnufied commented on pkg/diskmaker/controllers/lvset/reconcile.go:
```diff
@@ -165,6 +178,26 @@ DeviceLoop:
+		getTime := *r.deviceAgeMap.getTime
+
+		// skip devices younger than deviceMinAge, but remember them since they pass filters
+		if getTime().Sub(firstObserved) < deviceMinAge {
```
if we rename function above then, this could be written as  - `if currentTime.After(availableAt) {}` .

rohantmp replied:
how about this? also removes `getTime`:
```
// get/set firstObserved
		isOldEnough := r.deviceAgeMap.isOlderThan(blockDevice.KName, deviceMinAge)

// 		// skip devices younger than deviceMinAge, but remember them since they pass filters
		if isOldEnough {
```

gnufied commented on pkg/diskmaker/controllers/lvset/reconcile.go:
```diff
@@ -165,6 +178,26 @@ DeviceLoop:
 			}
 		}
+
+		getTime := *r.deviceAgeMap.getTime
```
I am not sure about - a function that just returns `time.Now()` leaking through here.

rohantmp commented:
addressed reviews, everything still works:
```
Events:
  Type    Reason               Age              From                               Message
  ----    ------               ----             ----                               -------
  Normal  DiscoveredNewDevice  4m (x6 over 4m)  localvolumeset-symlink-controller  minikube - found possible matching disk, waiting 1m0s to claim
  Normal  FoundMatchingDisk    3m (x3 over 3m)  localvolumeset-symlink-controller  minikube - symlinking matching disk
```

gnufied commented on pkg/diskmaker/controllers/lvset/reconcile.go:
```diff
@@ -165,6 +179,24 @@ DeviceLoop:
 		}
+
+		// skip devices younger than deviceMinAge
+		if isOldEnough {
```
but now this function and variable reads weird. It says - `if isOldEnough` which actually signals the opposite that the device is actually younger! can the function and variable be named appropriately?

rohantmp replied:
not sure I follow: `isOldEnough` signals that the device is older than `minDeviceAge`

rohantmp replied:
is it the comment?

rohantmp replied:
I get it, I'll do 
```
if !isOldEnough {
   continue DeviceLoop
}
```

gnufied commented on pkg/diskmaker/controllers/lvset/device_age.go:
```diff
@@ -0,0 +1,62 @@
+// checks if older than,
+// records current time if this is the first observation of key
+func (a *ageMap) isOlderThan(key string, t time.Duration) bool {
+
```
unnecessary new line.

gnufied commented on pkg/diskmaker/controllers/lvset/device_age.go:
```diff
@@ -0,0 +1,62 @@
+	firstObserved := a.getFirstObserved(key)
+	currentTime := *a.currentTime
+	return currentTime().Sub(firstObserved) > t
+
```
unnecessary new line.

gnufied commented on pkg/diskmaker/controllers/lvset/device_age.go:
```diff
@@ -0,0 +1,62 @@
+
+// getFirstObserved returns the age that the key was first observed and records it if
+// it was first observed now
+func (a *ageMap) getFirstObserved(key string) time.Time {
```
`getFirstObserved` is not merely getting the value, it is also storing and updating the map. IMO - a better name is needed for this function because a casual reader of code may think that this function is readonly..

gnufied commented on pkg/diskmaker/controllers/lvset/device_age_test.go:
```diff
@@ -0,0 +1,100 @@
+		targetLength := len(blockDevices) + increment
+		for i := len(blockDevices); i < targetLength; i++ {
+			blockDevices = append(blockDevices, internal.BlockDevice{KName: fmt.Sprintf("dev-%d", len(blockDevices))})
+
```
unnecessary new line.

gnufied commented on pkg/diskmaker/controllers/lvset/device_age.go:
```diff
@@ -0,0 +1,62 @@
+func newAgeMap() *ageMap {
+	f := func() time.Time {
+		return time.Now()
+	}
```
I think using a function pointer to set time for test purposes looks kind of weird. Not right now but may be we should consider defining mock interface for time and then using it.

gnufied commented on pkg/diskmaker/controllers/lvset/device_age.go:
```diff
@@ -0,0 +1,62 @@
+var (
+	// deviceMinAge is the minimum age for a device to be considered safe to claim
+	// otherwise, it could be a device that some other entity has attached and we have not claimed.
+	deviceMinAge = time.Minute
```
we can do this in a follow up but IMO this should be configurable via CLI option in diskmaker, so as certain deployments can run with different values potentially without requiring code change. This however may require propagating upto `LocalVolumeSet` object if this has to be configurable. We can do this in a follow up.

gnufied commented:
/lgtm
